### PR TITLE
ci: skip Github CI on branch pushes for forks

### PR DESCRIPTION
**Pull request by Sjors**
When a contributor maintains a fork of the repo, any pull request they make to their own fork, or to the main repository, will trigger two CI runs one for the branch push and one for the pull request.

After this PR when `SKIP_BRANCH_PUSH` is set, pushes made to git branches inside fork repositories will no longer trigger CI runs, unless the git branches are associated with PRs in the fork repository, or the main repository.

The same behaviour was added for Cirrus in e9bfbb5414ab14ca14d8edcfdf77f28c9ed67c33.

Note to maintainers: `SKIP_BRANCH_PUSH=true` needs to be set for the GUI repo to maintain existing behaviour.

### CODE REVIEW

**Comment by ryanofsky** [comment]
Mentioning the history of this change in the PR description is helpful, but it would be clearer if it said what the change is and what effects it has.

From code comments in this PR and https://github.com/bitcoin/bitcoin/pull/29274#issuecomment-2188840483, it seems like the main intended effect of this change is to prevent github tasks from running twice in PRs and causing the following output:

> Scherm­afbeelding 2024-06-25 om 14 48 34

But in the last thread where this change was discussed (https://github.com/bitcoin/bitcoin/pull/29274#discussion_r1457587805) a potential concern was that it would cause CI not to run on branch pushes within forks like Inquisition and Knots.

This seems like it could be a real problem, though maybe it is not a big deal for the forks to modify this file if they want CI to run on branches they are modifying anyway.

**Comment by Sjors** [comment]
@ryanofsky I updated the description.

In the earlier issue both @luke-jr (Knots) and @ajtowns (Inquisition) seemed to indicate that the change wouldn't be a problem for them. See https://github.com/bitcoin/bitcoin/pull/29274#issuecomment-1903768404. Let me know if that's incorrect or if you changed your mind.

> not a big deal for the forks to modify this file if they want CI to run on branches they are modifying anyway

So far it seems I'm the only one using this and it's not bothering any other project. But if we end up with some forks that need to modify this file and others that don't, then I don't think this PR should be merged. The extra runs are annoying, but if needed I can wait for a glorious future in which Github CI gets their act together.

**Review by ryanofsky — APPROVED** [review]
Code review ACK 3067cd55f0fc79a7fa6342ca0600472cabcf2690.

Thanks for the explanation, that all sounds good to me.

The new description is definitely better than the original but it is not actually saying what behavior will be after this PR. Would be helpful to say something like "After this PR, pushes made to git branches inside fork repositories will no longer trigger CI runs, unless the git branches are associated with PRs in the fork repository, or the main repository." (assuming that is correct)

**Comment by Sjors** [comment]
Thanks, added that (correct) sentence to the description.

**Comment by Sjors** [comment]
A use case that mentioned to me offline is that some people, before opening a PR to the main repo, like to push their branch and have CI check it. They then wait for CI on the branch to pass before opening a new PR. Their main reason for this is to reduce noise on the main repo.

It seems this PR breaks that use case. I suggest we reconsider it if and when Github adds support for custom ENV variables on a fork, like Cirrus does.

**Comment by willcl-ark** [comment]
From https://github.com/bitcoin/bitcoin/commit/e9bfbb5414ab14ca14d8edcfdf77f28c9ed67c33:

> When a contributor maintains a fork of the repo, any pull request they make
> to their own fork, or to the main repository, will trigger two CI runs:
> one for the branch push and one for the pull request.
> This can be avoided by setting SKIP_BRANCH_PUSH=true as a custom env variable
> in Cirrus repository settings, accessible from

GitHub supports setting custom env vars in each repo like Cirrus: https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/store-information-in-variables#defining-configuration-variables-for-multiple-workflows

**Comment by Sjors** [comment]
Oh wow that is remarkably well hidden! In that case I will change this PR to apply the same behaviour as Cirruss.

SKIP_BRANCH_PUSH=true

**Comment by vasild** [comment]
Concept ACK

> A use case that was mentioned to me offline is that some people, before opening a PR to the main repo, like to push their branch and have CI check it.

FWIW I do that but I have a dedicated dummy branch which is associated with a PR to merge it into my copy of `master`. So I never merge that PR and keep `push -f`ing into the dummy branch to test.

**Comment by Sjors** [comment]
Github documentation is pretty terrible. The change here should work, but the jobs are not skipped in https://github.com/Sjors/bitcoin/pull/70

**Comment by m3dwards** [comment]
Pretty sure this is what you are looking for:

```diff
diff --git a/.github/workflows/ci.yml b/.github/workflows/ci.yml
index cc1d8dd905..131f350111 100644
--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -85,7 +85,7 @@ jobs:
     # one for the branch push and one for the pull request.
     # This can be avoided by setting SKIP_BRANCH_PUSH=true as a custom env variable
     # in Github repository settings.
-    if: ${{ vars.SKIP_BRANCH_PUSH != 'true' }} || github.event_name == 'pull_request'
+    if: ${{ vars.SKIP_BRANCH_PUSH != 'true' || github.event_name == 'pull_request' }}

     timeout-minutes: 120

@@ -139,7 +139,7 @@ jobs:
     # See: https://github.com/actions/runner-images#available-images.
     runs-on: windows-2022

-    if: ${{ vars.SKIP_BRANCH_PUSH != 'true' }} || github.event_name == 'pull_request'
+    if: ${{ vars.SKIP_BRANCH_PUSH != 'true' || github.event_name == 'pull_request' }}

     env:
       PYTHONUTF8: 1
@@ -218,7 +218,7 @@ jobs:
   asan-lsan-ubsan-integer-no-depends-usdt:
     name: 'ASan + LSan + UBSan + integer, no depends, USDT'
     runs-on: ubuntu-24.04 # has to match container in ci/test/00_setup_env_native_asan.sh for tracing tools
-    if: ${{ vars.SKIP_BRANCH_PUSH != 'true' }} || github.event_name == 'pull_request'
+    if: ${{ vars.SKIP_BRANCH_PUSH != 'true' || github.event_name == 'pull_request' }}
     timeout-minutes: 120
     env:
       FILE_ENV: "./ci/test/00_setup_env_native_asan.sh"
```

**Comment by Sjors** [comment]
@m3dwards thanks! I'll give that a try. Any idea what was wrong with the syntax I used?

**Comment by Sjors** [comment]
That worked! See https://github.com/Sjors/bitcoin/pull/70/checks where the "push" actions are skipped.

**Comment by m3dwards** [comment]
> Any idea what was wrong with the syntax I used?

The `|| github.event_name == 'pull_request'` is code that needs to be executed rather than a static data so it needs to be inside the double quotes. I believe the hardcoded string of `|| github.event_name == 'pull_request'` is truthy hence it would run the job. Github Actions is finicky!

**Review by vasild — APPROVED** [review]
ACK 8610bcef9d030013f9e36cffe0c58dd2cfe85d66

**Comment by m3dwards** [comment]
ACK 8610bcef9d030013f9e36cffe0c58dd2cfe85d66

**Review by m3dwards — APPROVED** [review]

**Comment by Sjors** [comment]
Just repeating my "note to maintainers" from the description: like with Cirrus, `SKIP_BRANCH_PUSH=true` needs to be set for the GUI repo to maintain existing behaviour.

**Comment by achow101** [comment]
ACK 8610bcef9d030013f9e36cffe0c58dd2cfe85d66

**Pull request merged**
